upup/pkg/fi/cloudup/hetzner/cloud.go

```diff
@@ -0,0 +1,523 @@
+	panic("implement me")
+}
+
+// NewHetznerCloud returns a Cloud, using the env var HCLOUD_TOKEN
+func NewHetznerCloud(region string) (HetznerCloud, error) {
+	accessToken := os.Getenv("HCLOUD_TOKEN")
+	if accessToken == "" {
+		return nil, errors.New("HCLOUD_TOKEN is required")
+	}
+
+	opts := []hcloud.ClientOption{
+		hcloud.WithToken(accessToken),
+	}
+	client := hcloud.NewClient(opts...)
+
+	return &hetznerCloudImplementation{
+		Client: client,
+		dns:    nil,
+		region: region,
+	}, nil
+}
+
+// ProviderID returns the kops api identifier for Hetzner cloud provider
+func (c *hetznerCloudImplementation) ProviderID() kops.CloudProviderID {
+	return kops.CloudProviderHetzner
+}
+
+// Region returns the Hetzner region
+func (c *hetznerCloudImplementation) Region() string {
+	return c.region
+}
+
+func (c *hetznerCloudImplementation) GetCloudGroups(cluster *kops.Cluster, instancegroups []*kops.InstanceGroup, warnUnmatched bool, nodes []v1.Node) (map[string]*cloudinstances.CloudInstanceGroup, error) {
+	return getCloudGroups(c, cluster, instancegroups, warnUnmatched, nodes)
+}
+
+// DeleteGroup is not implemented
+func (c *hetznerCloudImplementation) DeleteGroup(g *cloudinstances.CloudInstanceGroup) error {
+	klog.V(8).Info("hetzner cloud provider DeleteGroup not implemented")
+	return fmt.Errorf("hetzner cloud provider does not support deleting cloud groups at this time")
+}
+
+// DeregisterInstance is not implemented
+func (c *hetznerCloudImplementation) DeregisterInstance(i *cloudinstances.CloudInstance) error {
+	klog.V(8).Info("hetzner cloud provider DeregisterInstance not implemented")
+	return nil
+}
+
+// FindVPCInfo is not implemented, it's only here to satisfy the fi.Cloud interface
+func (c *hetznerCloudImplementation) FindVPCInfo(id string) (*fi.VPCInfo, error) {
+	return nil, errors.New("hetzner cloud provider FindVPCInfo not implemented")
+}
+
+func (c *hetznerCloudImplementation) GetVPCUUID(networkCIDR string, vpcName string) (string, error) {
+	vpcUUID := ""
+	done, err := vfs.RetryWithBackoff(readBackoff, func() (bool, error) {
+		vpcs, err := c.GetAllVPCs()
+		if err != nil {
+			return false, err
+		}
+
+		for _, vpc := range vpcs {
+			if vpc.IPRange.Network() == networkCIDR && vpc.Name == vpcName {
+				vpcUUID = strconv.Itoa(vpc.ID)
+				return true, nil
+			}
+		}
+
+		return false, fmt.Errorf("vpc not yet created..")
+	})
+
+	if err != nil {
+		return "", err
+	}
+
+	if done {
+		return vpcUUID, nil
+	} else {
+		return "", wait.ErrWaitTimeout
+	}
+}
+
+func (c *hetznerCloudImplementation) GetApiIngressStatus(cluster *kops.Cluster) ([]fi.ApiIngressStatus, error) {
+	var ingresses []fi.ApiIngressStatus
+	done, err := vfs.RetryWithBackoff(readBackoff, func() (bool, error) {
+		if cluster.Spec.MasterPublicName != "" {
+			// Note that this must match Digital Ocean's lb name
+			klog.V(2).Infof("Querying Hetzner to find Loadbalancers for API (%q)", cluster.Name)
+
+			loadBalancers, err := c.GetAllLoadBalancers()
```

Comment:
Nit: looks like we can filter by name https://docs.hetzner.cloud/#load-balancers-get-all-load-balancers

Maybe add a TODO?